pageserver/src/consumption_metrics/upload.rs

```diff
@@ -0,0 +1,264 @@
+            );
+        }
+
+        let res = serde_json::to_writer(
+            (&mut buffer).writer(),
+            &EventChunk {
+                events: (&events[..chunk.len()]).into(),
+            },
+        );
+
+        match res {
+            Ok(()) => Some(Ok((chunk, buffer.split().freeze()))),
+            Err(e) => Some(Err(e)),
+        }
+    })
+}
+
+trait KeyGen<'a>: Copy {
+    fn generate(&self) -> IdempotencyKey<'a>;
+}
+
+impl<'a> KeyGen<'a> for &'a str {
+    fn generate(&self) -> IdempotencyKey<'a> {
+        IdempotencyKey::generate(self)
+    }
+}
+
+enum UploadError {
+    Rejected(reqwest::StatusCode),
+    Reqwest(reqwest::Error),
+    Cancelled,
+}
+
+impl std::fmt::Debug for UploadError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // use same impl because backoff::retry will log this using both
+        std::fmt::Display::fmt(self, f)
+    }
+}
+
+impl std::fmt::Display for UploadError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        use UploadError::*;
+
+        match self {
+            Rejected(code) => write!(f, "server rejected the metrics with {code}"),
+            Reqwest(e) => write!(f, "request failed: {e}"),
+            Cancelled => write!(f, "cancelled"),
+        }
+    }
+}
+
+impl UploadError {
+    fn is_reject(&self) -> bool {
+        matches!(self, UploadError::Rejected(_))
+    }
+}
+
+async fn upload(
+    client: &reqwest::Client,
+    metric_collection_endpoint: &reqwest::Url,
+    body: bytes::Bytes,
+    cancel: &CancellationToken,
+) -> Result<(), UploadError> {
+    if cancel.is_cancelled() {
+        // retry does not drop the request when cancel comes, so pre-check it.
+        //
+        // it's not terrible, because those come at shutdown when we will have very little time to
+        // act anyways.
+        Err(UploadError::Cancelled)
+    } else {
+        let warn_after = 3;
+        let max_attempts = 10;
+        utils::backoff::retry(
+            move || {
+                let body = body.clone();
+                async move {
+                    let res = client
+                        .post(metric_collection_endpoint.clone())
+                        .header(reqwest::header::CONTENT_TYPE, "application/json")
+                        .body(body)
+                        .send()
+                        .await;
+
+                    let res = res.and_then(|res| res.error_for_status());
+
+                    // 10 redirects are normally allowed, so we don't need worry about 3xx
+                    match res {
+                        Ok(_response) => Ok(()),
+                        Err(e) => {
+                            let status = e.status().filter(|s| s.is_client_error());
+                            if let Some(status) = status {
+                                // rejection used to be a thing when the server could reject a
```

Comment:
This seems worth logging a warning on: if we are getting 4xx then we probably have an auth problem or a pageserver bug, right?

Reply:
I need to check backoff::retry logging for permanent errors, I would expect that to be the same as "running out of attempts", which is a `warn!`.

Reply:
Hehe, permanent error is not logged. I'll add error logging.

Reply:
Next push d35ded163